Source/Particles/Collision/MCCProcess.cpp

```diff
@@ -0,0 +1,129 @@
+/* Copyright 2021 Modern Electron
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#include "MCCProcess.H"
+#include "WarpX.H"
+
+MCCProcess::MCCProcess (
+                        const std::string& scattering_process,
+                        const std::string& cross_section_file,
+                        const amrex::Real energy )
+    : m_type(parseProcessType(scattering_process))
+    , m_energy_penalty(energy)
+{
+    amrex::Print() << "Reading file " << cross_section_file << " for "
+                   << scattering_process << " scattering cross-sections.\n";
+
+    // read the cross-section data file into memory
+    readCrossSectionFile(cross_section_file, m_energies, m_sigmas);
+
+    init();
+}
+
+template <typename InputVector>
+MCCProcess::MCCProcess (
+                        const std::string& scattering_process,
+                        const InputVector&& energies,
+                        const InputVector&& sigmas,
+                        const amrex::Real energy )
+    : m_type(parseProcessType(scattering_process))
+    , m_energy_penalty(energy)
+{
+    using std::begin;
+    using std::end;
+    m_energies.insert(m_energies.begin(), begin(energies), end(energies));
+    m_sigmas  .insert(m_sigmas.begin(),   begin(sigmas),   end(sigmas));
+
+    init();
+}
+
+void*
+MCCProcess::operator new ( std::size_t count )
+{
+    return amrex::The_Managed_Arena()->alloc(count);
+}
+
+void
+MCCProcess::operator delete ( void* ptr )
+{
+    amrex::The_Managed_Arena()->free(ptr);
+}
+
+void
+MCCProcess::init()
+{
+    m_energies_data = m_energies.data();
+    m_sigmas_data = m_sigmas.data();
+
+    // save energy grid parameters for easy use
+    m_grid_size = m_energies.size();
+    m_energy_lo = m_energies[0];
+    m_energy_hi = m_energies[m_grid_size-1];
+    m_sigma_lo = m_sigmas[0];
+    m_sigma_hi = m_sigmas[m_grid_size-1];
+    m_dE = m_energies[1] - m_energies[0];
```

Comment:
In my experience, calculating m_dE this way can be problematic because of round off. The differences between energy values won't always necessarily be the same (if the energies don't have nice values). I would suggest this instead:
```suggestion
    m_dE = (m_energy_hi - m_energy_lo)/(m_grid_size - 1.);
```